File: Notificare.Geo.iOS.Binding.Bridge/Sources/Models/NotificareRegion.swift
import Foundation
import NotificareGeoKit

@objc
public class NotificareRegion : NSObject {
    @objc public let regionId: String
    @objc public let name: String
    @objc public let regionDescription: String?
    @objc public let referenceKey: String?
    @objc public let geometry: NotificareRegionGeometry
    @objc public let advancedGeometry: NotificareRegionAdvancedGeometry?
    @objc public let major: NSNumber? // Cannot represent Int? in Objective-C.
    @objc public let distance: Double
    @objc public let timeZone: String
    @objc public let timeZoneOffset: Double

    @objc public init(regionId: String, name: String, regionDescription: String?, referenceKey: String?, geometry: NotificareRegionGeometry, advancedGeometry: NotificareRegionAdvancedGeometry?, major: NSNumber?, distance: Double, timeZone: String, timeZoneOffset: Double) {
        self.regionId = regionId
        self.name = name
        self.regionDescription = regionDescription
        self.referenceKey = referenceKey
        self.geometry = geometry
        self.advancedGeometry = advancedGeometry
        self.major = major
        self.distance = distance
        self.timeZone = timeZone
        self.timeZoneOffset = timeZoneOffset
    }

    public convenience init(from region: NotificareGeoKit.NotificareRegion) {
        self.init(
            regionId: region.id,
            name: region.name,
            regionDescription: region.description,
            referenceKey: region.referenceKey,
            geometry: NotificareRegionGeometry(from: region.geometry),
            advancedGeometry: region.advancedGeometry.map { NotificareRegionAdvancedGeometry(from: $0) },
            major: region.major.map { NSNumber(integerLiteral: $0) },
            distance: region.distance,
            timeZone: region.timeZone,
            timeZoneOffset: Double(region.timeZoneOffset)
        )
    }
}

@objc
public class NotificareRegionGeometry : NSObject {
    @objc public let type: String
    @objc public let coordinate: NotificareRegionCoordinate

    @objc public init(type: String, coordinate: NotificareRegionCoordinate) {
        self.type = type
        self.coordinate = coordinate
    }

    public convenience init(from geometry: NotificareGeoKit.NotificareRegion.Geometry) {
        self.init(
            type: geometry.type,
            coordinate: NotificareRegionCoordinate(from: geometry.coordinate)
        )
    }
}

@objc
public class NotificareRegionAdvancedGeometry : NSObject {
    @objc public let type: String
    @objc public let coordinates: [NotificareRegionCoordinate]

    @objc public init(type: String, coordinates: [NotificareRegionCoordinate]) {
        self.type = type
        self.coordinates = coordinates
    }

    public convenience init(from advancedGeometry: NotificareGeoKit.NotificareRegion.AdvancedGeometry) {
        self.init(
            type: advancedGeometry.type,
            coordinates: advancedGeometry.coordinates.map { NotificareRegionCoordinate(from: $0) }
        )
    }
}

@objc
public class NotificareRegionCoordinate : NSObject {
    @objc public let latitude: Double
    @objc public let longitude: Double

    @objc public init(latitude: Double, longitude: Double) {
        self.latitude = latitude
        self.longitude = longitude
    }

    public convenience init(from coordinate: NotificareGeoKit.NotificareRegion.Coordinate) {
        self.init(
            latitude: coordinate.latitude,
            longitude: coordinate.longitude
        )
    }
}

File: Notificare.Geo.Core/Models/NotificareRegion.cs
namespace NotificareSdk.Geo.Core.Models;

public class NotificareRegion
{
    public string Id { get; }
    public string Name { get; }
    public string? Description { get; }
    public string? ReferenceKey { get; }
    public NotificareRegionGeometry Geometry { get; }
    public NotificareRegionAdvancedGeometry? AdvancedGeometry { get; }
    public int? Major { get; }
    public double Distance { get; }
    public string TimeZone { get; }
    public double TimeZoneOffset { get; }

    public NotificareRegion(string id, string name, string? description, string? referenceKey,
        NotificareRegionGeometry geometry, NotificareRegionAdvancedGeometry? advancedGeometry, int? major,
        double distance, string timeZone, double timeZoneOffset)
    {
        Id = id;
        Name = name;
        Description = description;
        ReferenceKey = referenceKey;
        Geometry = geometry;
        AdvancedGeometry = advancedGeometry;
        Major = major;
        Distance = distance;
        TimeZone = timeZone;
        TimeZoneOffset = timeZoneOffset;
    }
}

public class NotificareRegionGeometry
{
    public string Type { get; }
    public NotificareRegionCoordinate Coordinate { get; }

    public NotificareRegionGeometry(string type, NotificareRegionCoordinate coordinate)
    {
        Type = type;
        Coordinate = coordinate;
    }
}

public class NotificareRegionAdvancedGeometry
{
    public string Type { get; }
    public IList<NotificareRegionCoordinate> Coordinates { get; }

    public NotificareRegionAdvancedGeometry(string type, IList<NotificareRegionCoordinate> coordinates)
    {
        Type = type;
        Coordinates = coordinates;
    }
}

public class NotificareRegionCoordinate
{
    public double Latitude { get; }
    public double Longitude { get; }

    public NotificareRegionCoordinate(double latitude, double longitude)
    {
        Latitude = latitude;
        Longitude = longitude;
    }
}
